src/QsCompiler/QirGeneration/DebugInfoManager.cs

```diff
@@ -0,0 +1,575 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Quantum.QIR.Emission;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Ubiquity.NET.Llvm;
+using Ubiquity.NET.Llvm.DebugInfo;
+using Ubiquity.NET.Llvm.Instructions;
+using Ubiquity.NET.Llvm.Types;
+using Ubiquity.NET.Llvm.Values;
+
+namespace Microsoft.Quantum.QsCompiler.QIR
+{
+    using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
+
+    /// <summary>
+    /// This class holds shared utility routines for generating QIR debug information.
+    /// It contains a reference to the GenerationContext that owns it because it needs
+    /// access to the same shared state represented by the GenerationContext.
+    /// </summary>
+    internal sealed class DebugInfoManager
+    {
+        private static class Config
+        {
+            /// <summary>
+            /// Whether or not to emit debug information during QIR generation
+            /// </summary>
+            public static bool DebugSymbolsEnabled { get; set; } = false;
+
+            /// <summary>
+            /// Dwarf version we are using for the debug info in the QIR generation
+            /// </summary>
+            public static uint DwarfVersion { get; } = 4;
+
+            /// <summary>
+            /// Title of the CodeView module flag for debug info
+            /// </summary>
+            public static string CodeviewName { get; } = "CodeView";
+
+            /// <summary>
+            /// CodeView version we are using for the debug info in the QIR generation
+            /// </summary>
+            public static uint CodeviewVersion { get; } = 1;
+
+            /// <summary>
+            /// The source language information for Dwarf.
+            /// For now, we are using the C interface. Ideally this would be a user defined language for Q#.
+            /// </summary>
+            public static SourceLanguage QSharpLanguage { get; } = SourceLanguage.C99;
```

Comment:
Since this is defined in DwarfEnumerations.cs, it could be expanded to include Q# just as it was expanded to include C#.

Reply:
You should be able to use the space reserved for user-defined languages in DwarfEnumeration.cs. However, there seems to be some bug in the LLVM Binding APIs we use such that if you define a value in the user-defined section and pass it in to the creation of a DICompileUnit, it gets translated into some absurdly large and out-of-range integer by the time it's emitted to the IR. Ian helped me take a brief look into this and we didn't figure out the answer

Reply:
Ultimately, we will want to ask DWARF to add Q# to their language enumeration, so it seemed most reasonable to just keep using C99 in the meantime. I wrote aboout this in my internship notes, but I will create work requests in GitHub for things like this so that I can link them into the code and they're easy to find

Reply:
Thanks for the detailed answer, Ryan!